add OnceCell to shared-mutability (#107)
ferrous-systems/rust-training: training-slides/src/shared-mutability.md

}
```

Note:

As an in-depth example of the borrow checker's limitations, consider the [Splitting Borrows](https://doc.rust-lang.org/nomicon/borrow-splitting.html) idiom, which allows one to borrow different fields of the same struct with different mutability semantics:

```rust
struct Foo {
    a: i32,
    b: i32,
    c: i32,
}

let mut x = Foo {a: 0, b: 0, c: 0};
let a = &mut x.a;
let b = &mut x.b;
let c = &x.c;
*b += 1;
let c2 = &x.c;
*a += 10;
println!("{} {} {} {}", a, b, c, c2);
```

> The code works, but you *have* to fallback on special cases the borrow checker understands, like shadowing with `let a = &mut x.a;` or else the compiler will error. The borrow checker is particularly frail here - replacing `Foo` with `x = [1,2,3]` and trying to borrow indexes will make it error out.

Reviewer: ```suggestion
The code works, but once you have mutably borrowed a field, you cannot mutably borrow the whole value (e.g. by calling a method on it) at the same time - otherwise you could get two mutable references to the same field at the same time. 
```

}
```

Note:

As an in-depth example of the borrow checker's limitations, consider the [Splitting Borrows](https://doc.rust-lang.org/nomicon/borrow-splitting.html) idiom, which allows one to borrow different fields of the same struct with different mutability semantics:

```rust
struct Foo {
    a: i32,
    b: i32,
    c: i32,
}

let mut x = Foo {a: 0, b: 0, c: 0};
let a = &mut x.a;
let b = &mut x.b;
let c = &x.c;
*b += 1;
let c2 = &x.c;
*a += 10;
println!("{} {} {} {}", a, b, c, c2);
```

The code works, but you *have* to fallback on special cases the borrow checker understands, like shadowing with `let a = &mut x.a;` or else the compiler will error. The borrow checker is particularly frail here - replacing `Foo` with `x = [1,2,3]` and trying to borrow indexes will make it error out.

> Note:

Reviewer: I think we can remove this heading - we're already in a Note:

Reviewer: No, I meant you can remove the `Note:` because there's another `Note:` a few lines up and it doesn't make sense to have two notes on one slide.

}
```

Note:

> An advanced trainee may ask about the use of `Splitting Borrows` which would allow one to borrow different fields with different mutability semantics, such as

Reviewer: Speaker notes appear in the book version, and so should be written with that audience in mind.

Reviewer: @miguelraz any thoughts on how to re-word this to suit both the trainer and the trainee who is reading the book later on?

```rust
struct Foo {
    a: i32,
    b: i32,
    c: i32,
}

let mut x = Foo {a: 0, b: 0, c: 0};
let a = &mut x.a;
let b = &mut x.b;
let c = &x.c;
*b += 1;
let c2 = &x.c;
*a += 10;
println!("{} {} {} {}", a, b, c, c2);
```

> The code works, but you *have* to use shadowing with `let a = &mut x.a;` or else the compiler will error. The borrowchecker is particularly frail here - replacing `Foo` with `x = [1,2,3]` and trying to borrow indexes will make it error out.

Reviewer: I don't know that "You *have* to use shadowing" is true here. This code works too:

```rust
    let mut x = Foo {a: 0, b: 0, c: 0};
    let c = &x.c;
    x.b += 1;
    let c2 = &x.c;
    x.a += 10;
    println!("{} {} {} {}", x.a, x.b, c, c2);
```

The main take-away I think we want here is that the borrow checker is special-cased for borrowing fields in structs and tuples, and the magic there does not apply to borrowing elements using the indexing operator, like with arrays or hashmaps:

This also works:

```rust
    let mut z = (1, 2);
    let r = &z.1;
    z.0 += 1;
    println!("{:?}, {}", z, r);
```

This does not:

```rust
    let mut z = [1, 2];
    let r = &z[1];
    z[0] += 1;
    println!("{:?}, {}", z, r);
```

Reviewer: I have re-worded this

* `Rc<RefCell<T>>`
* (Multi-threaded programs might use `Arc<Mutex<T>>`)

## `OnceCell` for special cases

If you only need to modify a field *once*, a `OnceCell` can help you keep the ownership system checks at compile-time

```rust
fn main() {
    let post = Post {
        content: String::from("Blah"),
        ..Post::default()
    };

>     assert!(post.days_on_hn_front_page.get().is_none());

Reviewer: I'm not sure this works as a motivating example, because "days on HN front page" is a property that changes over time.

Perhaps "first viewed at", with a `std::time::Instant` might work better?

* `Rc<RefCell<T>>`
* (Multi-threaded programs might use `Arc<Mutex<T>>`)

## `OnceCell` for special cases

If you only need to modify a field *once*, a `OnceCell` can help you keep the ownership system checks at compile-time

```rust
use std::time::Instant;

fn main() {
    let post = Post {
        content: String::from("Blah"),
        ..Post::default()
    };
    assert!(post.first_viewed_at.get().is_none());
    println!("{:?}", post.date_of_first_view());
    assert!(post.first_viewed_at.get().is_some());
}

#[derive(Debug, Default)]
struct Post {
    content: String,
    first_viewed_at: std::cell::OnceCell<Instant>,
}

impl Post {
    fn date_of_first_view(&self) -> Instant {

>         *self.first_viewed_at.get_or_init(|| {Instant::now()})

Reviewer: ```suggestion
        *self.first_viewed_at.get_or_init(Instant::now)
```

`Instant::now` is a function with no arguments, which is exactly what get_or_init wants.

* `Rc<RefCell<T>>`
* (Multi-threaded programs might use `Arc<Mutex<T>>`)

## `OnceCell` for special cases

> If you only need to modify a field *once*, a `OnceCell` can help you keep the ownership system checks at compile-time

Reviewer: Can we link to the std-library docs on `OnceCell`?

```rust
use std::time::Instant;

fn main() {
    let post = Post {
        content: String::from("Blah"),
        ..Post::default()
    };
    assert!(post.first_viewed_at.get().is_none());
    println!("{:?}", post.date_of_first_view());
    assert!(post.first_viewed_at.get().is_some());
}

#[derive(Debug, Default)]
struct Post {
    content: String,
    first_viewed_at: std::cell::OnceCell<Instant>,
}

impl Post {
    fn date_of_first_view(&self) -> Instant {
        *self.first_viewed_at.get_or_init(Instant::now)
    }
}
```

> A [LazyCell](https://doc.rust-lang.org/std/cell/struct.LazyCell.html) will do this initialization lazily, and a [LazyLock](https://doc.rust-lang.org/std/sync/struct.LazyLock.html) will do it in a threadsafe way.

Reviewer: I'm afraid I didn't understand the difference between OnceCell and LazyCell after reading this note. I also observe that OnceLock exists.

The primary difference is that a LazyCell owns its initialisation function - it is passed when the cell is constructed, not when the cell is accessed. See https://doc.rust-lang.org/std/cell/index.html#oncecellt and https://doc.rust-lang.org/std/cell/index.html#lazycellt-f. However both are equally 'lazy' in that the function is not called at construction time - it is only called on first access (either explicitly or implicitly).

* `Rc<RefCell<T>>`
* (Multi-threaded programs might use `Arc<Mutex<T>>`)

## `OnceCell` for special cases

> If you only need to modify a field *once*, a [OnceCell](https://doc.rust-lang.org/stable/std/cell/struct.OnceCell.html) can help you keep the ownership system checks at compile-time.

Reviewer: ```suggestion
If you only need to initialise a field and never modify it, and you are not sharing the value across threads, a [OnceCell](https://doc.rust-lang.org/stable/std/cell/struct.OnceCell.html) is cheaper than a `RefCell`.
```
